### ISSUE TITLE
Multi adapter weight conflicts when services are concurrent

### ISSUE THREAD

**Issue by Nipi64310** [huggingface/peft#804]
### Feature request

Peft's set_adapter function is equal to a global mode setting. Weight conflicts arise  when I use multi adapter to construct two demos with different functionality, but someone plays both demos at the same time. Process A, for example, uses Lora A weights to calculate to layer 10. At this point, process B changes the active_adapter of the current model to lora B, forcing the 11th to final layer of process A to use lora B for calculation.

I can use accelerator.wait_for_everyone() to prevent this scenario during training, but I'm not sure whether there is a better approach to avoid conflicts created by this multi-process while reasoning. The current workaround is to add an adapter_name parameter to peft's lora forward function. The lora weight corresponding to the active_adapter will not be utilized if this parameter is supplied, however such a fix requires changing the appropriate model file (modeling_xxxx.py) to take effect.

### Motivation

#518

### Your contribution

If adding parameters works or if someone has a better idea, I can submit a PR to take care of this.

**Comment by BenjaminBossan** [comment]
So I assume that you're serving the model via some web server. It sounds like the solution would have to be applied on that level, by applying some kind of lock on access to the model while it is still processing the output or switching the adapter.

**Comment by Nipi64310** [comment]
> 

Yes, I use a webui-like service, but if you use the process LOCK, each process must acquire the lock in order to execute the model's inference, hence the service cannot be parallelized.

**Comment by BenjaminBossan** [comment]
If the service is using true multiprocessing, each process should have a copy of the model (in fact, of the whole Python program), so it would be strange if the error you described occurred there. It sounds more like concurrent usage of the same object (multi-threading) is happening here. If that's the case, locking seems like the best solution, but as you said it means blocking for other users.

**Comment by Nipi64310** [comment]
Thank you for responding. The issue I encountered is indeed related to the concurrency situation you mentioned. Is there a better solution than locking, because I don't want to block? The current method is to pass the adapter name parameter in the forward function but this necessitates modifying the modeling_xxxx.py file, which is ugly and not universal.

**Comment by BenjaminBossan** [comment]
> The current method is to pass the adapter name parameter in the forward function but this necessitates modifying the modeling_xxxx.py file, which is ugly and not universal.

You mean so that the model can decide on the fly what adapter to use based on the passed adapter name? IMO, that doesn't sound like such a bad solution. Yes, it is more work, but if it achieves the desired result, it doesn't sound too bad.

Alternatively, you could use true multiprocessing, with a sufficient number of processes to handle the number of concurrent users you have (horizontal scaling). But that means the memory requirement for your app is multiplied by the number of processes, so it can get quite expensive.

**Comment by Nipi64310** [comment]
Yes, the true multiprocessing consumes too many resources, so we can only do this for now and hope for a better solution in the future.

> submit

Hi, @BenjaminBossan ,
Yes, the true multiprocessing consumes too many resources, so we can only do this for now and hope for a better solution in the future.
I'm not sure if  need to submit a PR to add a passing adatper name parameter to each lora related class, even though this isn't directly available, but someone might need it.

**Comment by BenjaminBossan** [comment]
> I'm not sure if need to submit a PR to add a passing adatper name parameter to each lora related class, even though this isn't directly available, but someone might need it.

I'm not sure how exactly you solved it, but probably it's not something that we want to add generally, like changing the signature of `forward` for everyone using LoRA. However, I do think it would be useful for you to post the code here, so that users with similar issues could find it in the future. Maybe we can even figure out some improvements if you do.

**Comment by Nipi64310** [comment]
ok, This parameter is simple to add, for example, to modify the linear layer.

```python
def forward(self, x: torch.Tensor, adapter_name: str = None):
    previous_dtype = x.dtype
    # ----- 
    if self.merged == False and adapter_name is not None and adapter_name in self.lora_A.keys():
        result = F.linear(x, transpose(self.weight, self.fan_in_fan_out), bias=self.bias)
        x = x.to(self.lora_A[adapter_name].weight.dtype)

        result += (
            self.lora_B[adapter_name](
                self.lora_A[adapter_name](self.lora_dropout[adapter_name](x))
            )
            * self.scaling[adapter_name]
        )
    # ----- 
    else:
        if self.active_adapter not in self.lora_A.keys():
            return F.linear(x, transpose(self.weight, self.fan_in_fan_out), bias=self.bias)
        if self.disable_adapters:
            if self.r[self.active_adapter] > 0 and self.merged:
                self.unmerge()
            result = F.linear(x, transpose(self.weight, self.fan_in_fan_out), bias=self.bias)
        elif self.r[self.active_adapter] > 0 and not self.merged:
            result = F.linear(x, transpose(self.weight, self.fan_in_fan_out), bias=self.bias)

            x = x.to(self.lora_A[self.active_adapter].weight.dtype)

            result += (
                self.lora_B[self.active_adapter](
                    self.lora_A[self.active_adapter](self.lora_dropout[self.active_adapter](x))
                )
                * self.scaling[self.active_adapter]
            )
        else:
            result = F.linear(x, transpose(self.weight, self.fan_in_fan_out), bias=self.bias)

    result = result.to(previous_dtype)

    return result

from peft.tuners.lora import Linear
Linear.forward = forward

```

To make model.generate handle the adapter_name parameter, add it to the forward function of xxxForCausalLM, then pass it to the Lora layer corresponding to xxxAttention, and then modify the linear parameter passing. eg: [bloom](https://github.com/huggingface/transformers/blob/main/src/transformers/models/bloom/modeling_bloom.py#L290)

```python
#fused_qkv = self.query_key_value(hidden_states)
#The original code above has been modified as follows:
    
if isinstance(self.query_key_value,LoraLayer):
	fused_qkv = self.query_key_value(hidden_states,adapter_name=adapter_name)
else:
    fused_qkv = self.query_key_value(hidden_states)
```

**Comment by Nipi64310** [comment]
> Thanks @Nipi64310. One thing I wonder about is that the code you provide does not appear to do anything too different from what would happen if `set_adapter` is called. Still, as you report, it works, whereas `set_adapter` does not. I'm not sure what the cause is, if I had to guess it's probably that the loop in `set_adapter` is not atomic:
> 
> https://github.com/huggingface/peft/blob/ed396a69ed6469be87f90e739f98f19ec9973983/src/peft/tuners/lora.py#L513-L518
> 
> Regarding a more general adoption of your proposal: I could see adding an optional `adapter_name: Optional[str]` argument to all the lora `forward` methods, and then inside the methods, do `active_adapter = self.active_adapter if adapter_name is None else adapter_name`. However, the problem, as you describe, is still that the caller needs to pass additional arguments down, which we cannot guarantee (it seems that it's not generally true for transformers models).
> 
> ping @pacman100 WDYT?

The code just replied is really crude, just to make an example. In addition, if the adapter_name is provided when model inference, I think set_adapter should not be important, I am looking at the source code of PEFT, can there be a better way to deal with it, if so, I will paste the available code.

**Comment by aaajun-czb** [comment]
> > Thanks @Nipi64310. One thing I wonder about is that the code you provide does not appear to do anything too different from what would happen if `set_adapter` is called. Still, as you report, it works, whereas `set_adapter` does not. I'm not sure what the cause is, if I had to guess it's probably that the loop in `set_adapter` is not atomic:
> > https://github.com/huggingface/peft/blob/ed396a69ed6469be87f90e739f98f19ec9973983/src/peft/tuners/lora.py#L513-L518
> > 
> > Regarding a more general adoption of your proposal: I could see adding an optional `adapter_name: Optional[str]` argument to all the lora `forward` methods, and then inside the methods, do `active_adapter = self.active_adapter if adapter_name is None else adapter_name`. However, the problem, as you describe, is still that the caller needs to pass additional arguments down, which we cannot guarantee (it seems that it's not generally true for transformers models).
> > ping @pacman100 WDYT?
> 
> The code just replied is really crude, just to make an example. In addition, if the adapter_name is provided when model inference, I think set_adapter should not be important, I am looking at the source code of PEFT, can there be a better way to deal with it, if so, I will paste the available code.

I am currently preparing to launch a project to parallelize the operation of multiple PEFT models under insufficient device resources. My assumption is that it is possible to load a pre trained model on a single card, and then load the parameters of multiple downstream tasks' PEFT simultaneously (meaning that the graphics memory>(pre trained model and all parameters)).
Then, when inferring multiple models based on the same pre trained model in parallel, it will automatically guide its PEFT parameters.

My idea comes from the following paper: PetS: A Unified Framework for Parameter Efficient Transformers Serving

I think your work requirements for parallelizing reasoning are very similar to this one, aiming to improve the utilization of GPU. I believe that simply modifying the source code of PEFT may not be able to complete this task. The author of the paper has customized the inference related interface through the CUDA interface, and perhaps should consider writing the relevant library themselves. But I am a beginner in CUDA programming. If there is any research progress, please feel free to communicate.

**Comment by Nipi64310** [comment]
Hi @aaajun-czb ,
I don’t know if this [[repository](https://github.com/sabetAI/BLoRA)](https://github.com/sabetAI/BLoRA) is helpful to you. I have solved my problem by modifying vllm.

**Comment by aaajun-czb** [comment]
> Hi @aaajun-czb , I don’t know if this [[repository](https://github.com/sabetAI/BLoRA)](https://github.com/sabetAI/BLoRA) is helpful to you. I have solved my problem by modifying vllm.

Thank you! It really give me some idea about how to do it.

As my paper deadline is 11.30, I may share my code near that day, after my work done.

I will focus on how to support the parallel inference for multi peft models. Hope my work will give someone help.



> Note that we're also currently working on supporting mixed LoRA batches in #903. It's not exactly the same thing as OP is asking for, but could still provide a viable solution.

I admit that I'm not very familiar with structure of PEFT, but my work may combine it. I will pay attention to the issue.

**Comment by Nipi64310** [comment]
> Note that we're also currently working on supporting mixed LoRA batches in #903. It's not exactly the same thing as OP is asking for, but could still provide a viable solution.

This is good and can already solve the problem, but the implementation of PR may significantly slow down the inference speed.

**Comment by BenjaminBossan** [comment]
Mixing adapters within a batch is expected to be slower than having only a single adapter per batch. If you have some proposal to speed up the proposed implementation, please let us know.

**Comment by Nipi64310** [comment]
> Mixing adapters within a batch is expected to be slower than having only a single adapter per batch. If you have some proposal to speed up the proposed implementation, please let us know.

Sorry, I'm not sure how to optimize, but I think the tensor[mask] operation should be avoided because it is very slow.

> ```python
> sub_batch = x[sub_batch_indices_list[i]].to(lora_A.weight.dtype)
> lora_output = lora_B(lora_A(dropout(sub_batch))) * scaling
> final_output[sub_batch_indices_list[i]] += lora_output.to(previous_dtype)
> ```

**Comment by BenjaminBossan** [comment]
From internal testing, we know that the suggested implementation will be faster than having batch size 1 and switching adapters between each sample. So there is some progress, even though there might be room for speed ups. If we, or someone else, comes up with a better suggestion, we're happy to make the change.

**Comment by whybeyoung** [comment]
can re open this ?

**Comment by codernew007** [comment]
saw solution in [s-lora](https://github.com/S-LoRA/S-LoRA), but not all huggingface models are supported

**Comment by sunjunlishi** [comment]
@Nipi64310 does peft support vllm speedup
does it surrpost vllm to realize "model = PeftModel.from_pretrained(model, adapter_to_resume, is_trainable=False)"

**Comment by BenjaminBossan** [comment]
> Note that we're also currently working on supporting mixed LoRA batches in #903. It's not exactly the same thing as OP is asking for, but could still provide a viable solution.

We merged #1558, a follow up to #903. If you want to give this a try and report back if the performance is good for you, that would be great.
